### PR TITLE
Make `optparse-applicative` forward-`-Wcompat`ible

### PR DESCRIPTION
**Pull request by hvr**
See also
https://ghc.haskell.org/trac/ghc/wiki/Migration/8.0#base-4.9.0.0

### CODE REVIEW

File: optparse-applicative.cabal [hunk 1]
```diff
@@ -51,10 +51,11 @@ source-repository head
...
 library
-  if impl(ghc >= 8)
-    ghc-options:       -Wall -Wno-redundant-constraints -Wcompat
-  else
-    ghc-options:       -Wall
+  ghc-options:         -Wall
+
+  -- See https://ghc.haskell.org/trac/ghc/wiki/Migration/8.0#base-4.9.0.0
+  if impl(ghc >= 8.0)
+    ghc-options:  -Wno-redundant-constraints -Wcompat -Wnoncanonical-monad-instances -Wnoncanonical-monadfail-instances
```
Comment (HuwCampbell) [inline optparse-applicative.cabal]: Is it ok to have two active `ghc-options` in scope like this? Will `Wall` still be on for ghc 8?

**Comment by HuwCampbell** [comment]
Looks good, I just managed to get compat working for the remaining semigroup classes a few weeks ago, so I'm happy to future proof further.

**Comment by hvr** [comment]
@HuwCampbell I actually had a patch sitting here which included the semigroup classes... but I forgot to make it into a PR... so this PR is actually a stripped down version of that :-)

File: optparse-applicative.cabal (hunk 1 above)
Comment (hvr) [inline optparse-applicative.cabal]: `ghc-options` are combined via a concatenation monoid. That generally applies for all cabal  fields which are list-ish (such as options or `other-modules` or `build-depends` etc).

**Comment by HuwCampbell** [comment]
Thanks

**Pull request merged**